Geometry/MTDGeometryBuilder/src/MTDGeometry.cc

```diff
@@ -0,0 +1,257 @@
+#include <typeinfo>
+
+#include "Geometry/MTDGeometryBuilder/interface/MTDGeometry.h"
+#include "Geometry/MTDNumberingBuilder/interface/GeometricTimingDet.h"
+#include "Geometry/CommonDetUnit/interface/GeomDet.h"
+#include "Geometry/CommonDetUnit/interface/GeomDetType.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "DataFormats/ForwardDetId/interface/MTDDetId.h"
+#include "DataFormats/ForwardDetId/interface/BTLDetId.h"
+#include "DataFormats/ForwardDetId/interface/ETLDetId.h"
+
+#include "DataFormats/DetId/interface/DetId.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+
+namespace {
+  GeomDetEnumerators::SubDetector
+  geometricDetToGeomDet(GeometricTimingDet::GTDEnumType gdenum) {
+    // provide a map between the GeometricTimingDet enumerators and the GeomDet enumerators of the possible tracker subdetectors
+    if(gdenum == GeometricTimingDet::GTDEnumType::BTL ) return GeomDetEnumerators::SubDetector::TimingBarrel;
+    if(gdenum == GeometricTimingDet::GTDEnumType::ETL ) return GeomDetEnumerators::SubDetector::TimingEndcap;    
+    return GeomDetEnumerators::SubDetector::invalidDet;
+  }
+  
+  class DetIdComparator {
+  public:
+    bool operator()(GeometricTimingDet const* gd1, GeometricTimingDet const * gd2) const {
+      uint32_t det1 = gd1->geographicalId();
+      uint32_t det2 = gd2->geographicalId();
+      return det1 < det2;
+    }  
+  };
+}
+
+MTDGeometry::MTDGeometry(GeometricTimingDet const* gd)
+   : theTrackerDet(gd)
+{
+  for(unsigned int i=0;i<2;++i) {
+    theSubDetTypeMap[i] = GeomDetEnumerators::invalidDet;
+    theNumberOfLayers[i] = 0;
+  }
+  GeometricTimingDet::ConstGeometricTimingDetContainer subdetgd = gd->components();
+  
+  LogDebug("BuildingSubDetTypeMap") 
+    << "GeometricTimingDet and GeomDetEnumerators enumerator values of the subdetectors" << std::endl;
+  for(unsigned int i=0;i<subdetgd.size();++i) {
+    MTDDetId mtdid(subdetgd[i]->geographicalId());
+    assert(mtdid.mtdSubDetector()>0 && mtdid.mtdSubDetector()<3);
+    theSubDetTypeMap[mtdid.mtdSubDetector()-1] = geometricDetToGeomDet(subdetgd[i]->type());
+    theNumberOfLayers[mtdid.mtdSubDetector()-1]= subdetgd[i]->components().size();   
+    LogTrace("BuildingSubDetTypeMap") 
+     << "subdet " << i 
+     << " Geometric Det type " << subdetgd[i]->type()
+     << " Geom Det type " << theSubDetTypeMap[mtdid.mtdSubDetector()-1]
+     << " detid " << std::hex <<  subdetgd[i]->geographicalId().rawId() << std::dec
+     << " subdetid " <<  mtdid.mtdSubDetector()
+     << " number of layers " << subdetgd[i]->components().size()
+     << std::endl;
+  }
+  LogDebug("SubDetTypeMapContent") 
```

Comment:
Better to use the new `log` function of the message logger. You pass it a lambda which will only be evaluated if the message will be printed.

Reply:
done